lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsInput.java

```diff
@@ -0,0 +1,125 @@
+import io.trino.filesystem.TrinoInput;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.OptionalLong;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.filesystem.gcs.GcsUtils.getReadChannel;
+import static io.trino.filesystem.gcs.GcsUtils.handleGcsException;
+import static java.util.Objects.checkFromIndexSize;
+import static java.util.Objects.requireNonNull;
+
+final class GcsInput
+        implements TrinoInput
+{
+    private final GcsLocation location;
+    private final Blob blob;
+    private final int readBlockSize;
+    private OptionalLong length;
+    private boolean closed;
+
+    // Note: default read chunk size is 2Mib - see com.google.cloud.storage.BaseStorageReadChannel
+    public GcsInput(GcsLocation location, Blob blob, int readBlockSize, OptionalLong length)
+    {
+        this.location = requireNonNull(location, "location is null");
+        this.blob = requireNonNull(blob, "storage is null");
+        checkArgument(readBlockSize >= 0, "readBlockSize is negative");
+        this.readBlockSize = readBlockSize;
+        this.length = requireNonNull(length, "length is null");
+    }
+
+    @Override
+    public void readFully(long position, byte[] buffer, int bufferOffset, int bufferLength)
+            throws IOException
+    {
+        ensureOpen();
+        if (position < 0) {
+            throw new IOException("Negative seek offset");
+        }
+        checkFromIndexSize(bufferOffset, bufferLength, buffer.length);
+        if (bufferLength == 0) {
+            return;
+        }
+
+        try (ReadChannel readChannel = getReadChannel(blob, location, position, readBlockSize, length)) {
+            int readSize = readNBytes(readChannel, buffer, bufferOffset, bufferLength);
+            if (readSize != bufferLength) {
+                throw new EOFException("End of file reached before reading fully: " + location);
+            }
+        }
+        catch (RuntimeException e) {
+            throw handleGcsException(e, "reading file", location);
+        }
+    }
+
+    private int readNBytes(ReadChannel readChannel, byte[] buffer, int bufferOffset, int bufferLength)
+            throws IOException
+    {
+        ByteBuffer wrappedBuffer = ByteBuffer.wrap(buffer, bufferOffset, bufferLength);
+        int readSize = 0;
+        while (readSize < bufferLength) {
+            int bytesRead = readChannel.read(wrappedBuffer);
+            if (bytesRead == -1) {
+                break;
+            }
+            readSize += bytesRead;
+        }
+        return readSize;
+    }
+
+    @Override
+    public int readTail(byte[] buffer, int bufferOffset, int bufferLength)
+            throws IOException
+    {
+        ensureOpen();
+        checkFromIndexSize(bufferOffset, bufferLength, buffer.length);
+        try (ReadChannel readChannel = getReadChannel(blob, location, blob.getSize() - bufferLength, readBlockSize, length)) {
```

Comment:
This computation could result in a negative offset, since it's legal to do a tail read that is larger than the file size, and in some cases the clients don't know the actual size. Instead, we should use the `Range` header to read from the tail of the file, like we do for S3: https://cloud.google.com/storage/docs/xml-api/reference-headers#range

Reply:
Checked the gcs code: the seek method sets the range header using the ByteRangeSpec internally (StorageReadChannel, BlobReadChannelV2), the api is package private. Changed to min(0, size - bufferlength), lmk if that works.